File: backend/IoTTeamCenter.Api/Infrastructure/ScheduleCalculator.cs
namespace IoTTeamCenter.Api.Infrastructure;

public sealed record ScheduleCalculationTask(
    long Id,
    long? ParentId,
    int SortOrder,
    DateOnly? PlanStart,
    int PlanDays,
    string StartMode,
    long? PredecessorId,
    int LagDays,
    DateOnly? ActualStart,
    DateOnly? ActualFinish,
    DateOnly? ForecastFinish,
    decimal PercentComplete,
    string Status);

public sealed class ResolvedScheduleTask
{
    public required ScheduleCalculationTask Source { get; init; }
    public required string Wbs { get; set; }
    public required int Depth { get; set; }
    public DateOnly? PlanStart { get; init; }
    public DateOnly? PlanFinish { get; init; }
    public DateOnly? ActualStart { get; init; }
    public DateOnly? ActualFinish { get; init; }
    public DateOnly? ForecastFinish { get; init; }
    public required int WorkDays { get; init; }
    public required int Weight { get; init; }
    public required decimal PercentComplete { get; init; }
    public required string Status { get; init; }
    public required IReadOnlyList<ResolvedScheduleTask> Children { get; init; }
}

public sealed record ResolvedSchedule(
    IReadOnlyList<ResolvedScheduleTask> Roots,
    IReadOnlyDictionary<long, ResolvedScheduleTask> ById);

/// <summary>
/// Resolves linked dates and WBS roll-ups from immutable database facts. No
/// derived parent value is accepted from a client or persisted during normal
/// schedule edits.
/// </summary>
public static class ScheduleCalculator
{
    public static ResolvedSchedule Resolve(
        IReadOnlyCollection<ScheduleCalculationTask> tasks,
        IReadOnlySet<DateOnly> holidays)
    {
        var byId = tasks.ToDictionary(task => task.Id);
        var children = tasks
            .GroupBy(task => task.ParentId ?? 0)
            .ToDictionary(
                group => group.Key,
                group => group.OrderBy(task => task.SortOrder).ThenBy(task => task.Id).ToArray());
        var resolved = new Dictionary<long, ResolvedScheduleTask>();
        var visiting = new HashSet<long>();

        ResolvedScheduleTask ResolveOne(ScheduleCalculationTask task)
        {
            if (resolved.TryGetValue(task.Id, out var cached)) return cached;
            if (!visiting.Add(task.Id))
                throw new ApiException(StatusCodes.Status409Conflict, "schedule_cycle", "The schedule contains a circular parent or predecessor relationship.");

            var resolvedChildren = children.GetValueOrDefault(task.Id, [])
                .Select(ResolveOne)
                .ToArray();

            ResolvedScheduleTask result;
            if (resolvedChildren.Length > 0)
            {
                var planStart = Minimum(resolvedChildren.Select(child => child.PlanStart));
                var planFinish = Maximum(resolvedChildren.Select(child => child.PlanFinish));
                var actualStart = Minimum(resolvedChildren.Select(child => child.ActualStart));
                var actualFinish = resolvedChildren.All(child => child.Status == "Done")
                    ? Maximum(resolvedChildren.Select(child => child.ActualFinish))
                    : null;
                var forecastFinish = Maximum(resolvedChildren.Select(child => child.ForecastFinish ?? child.PlanFinish));
                var weight = resolvedChildren.Sum(child => child.Weight);
                var weightedProgress = weight == 0
                    ? 0
                    : resolvedChildren.Sum(child => child.PercentComplete * child.Weight) / weight;
                var status = resolvedChildren.All(child => child.Status == "Done")
                    ? "Done"
                    : resolvedChildren.Any(child => child.Status == "Blocked")
                        ? "Blocked"
                        : resolvedChildren.Any(child => child.Status != "Not Started" || child.PercentComplete > 0 || child.ActualStart is not null)
                            ? "In Progress"
                            : "Not Started";

                result = new ResolvedScheduleTask
                {
                    Source = task,
                    Wbs = string.Empty,
                    Depth = 0,
                    PlanStart = planStart,
                    PlanFinish = planFinish,
                    ActualStart = actualStart,
                    ActualFinish = actualFinish,
                    ForecastFinish = forecastFinish,
                    WorkDays = NetworkDays(planStart, planFinish, holidays),
                    Weight = Math.Max(1, weight),
                    PercentComplete = decimal.Round(weightedProgress, 2, MidpointRounding.AwayFromZero),
                    Status = status,
                    Children = resolvedChildren
                };
            }
            else
            {
                var planStart = task.PlanStart;
                if (task.StartMode == "linked" && task.PredecessorId is long predecessorId)
                {
                    if (!byId.TryGetValue(predecessorId, out var predecessor))
                        throw new ApiException(StatusCodes.Status409Conflict, "schedule_predecessor_missing", "A linked task refers to an inactive predecessor.");
                    var predecessorFinish = ResolveOne(predecessor).PlanFinish;
                    if (predecessorFinish is not null)
                        planStart = NextWorkDay(SafeAddDays(predecessorFinish.Value, 1 + task.LagDays), holidays);
                }

                DateOnly? planFinish = planStart is null
                    ? null
                    : SafeAddDays(planStart.Value, Math.Max(1, task.PlanDays) - 1);
                var workDays = NetworkDays(planStart, planFinish, holidays);
                result = new ResolvedScheduleTask
                {
                    Source = task,
                    Wbs = string.Empty,
                    Depth = 0,
                    PlanStart = planStart,
                    PlanFinish = planFinish,
                    ActualStart = task.ActualStart,
                    ActualFinish = task.ActualFinish,
                    ForecastFinish = task.ForecastFinish,
                    WorkDays = workDays,
                    Weight = Math.Max(1, workDays),
                    PercentComplete = task.PercentComplete,
                    Status = task.Status,
                    Children = []
                };
            }

            visiting.Remove(task.Id);
            resolved.Add(task.Id, result);
            return result;
        }

        var roots = children.GetValueOrDefault(0, []).Select(ResolveOne).ToArray();
        foreach (var task in tasks) ResolveOne(task);

        void Number(IReadOnlyList<ResolvedScheduleTask> siblings, string prefix, int depth)
        {
            for (var index = 0; index < siblings.Count; index++)
            {
                var item = siblings[index];
                item.Wbs = string.IsNullOrEmpty(prefix) ? $"{index + 1}" : $"{prefix}.{index + 1}";
                item.Depth = depth;
                Number(item.Children, item.Wbs, depth + 1);
            }
        }

        Number(roots, string.Empty, 0);
        return new ResolvedSchedule(roots, resolved);
    }

    public static int NetworkDays(DateOnly? start, DateOnly? finish, IReadOnlySet<DateOnly> holidays)
    {
        if (start is null || finish is null || finish < start) return 0;
        var count = 0;
        for (var day = start.Value; ; day = SafeAddDays(day, 1))
        {
            if (IsWorkDay(day, holidays)) count++;
            if (day == finish.Value) break;
        }
        return count;
    }

    private static DateOnly NextWorkDay(DateOnly date, IReadOnlySet<DateOnly> holidays)
    {
        var candidate = date;
        while (!IsWorkDay(candidate, holidays)) candidate = SafeAddDays(candidate, 1);
        return candidate;
    }

    private static DateOnly SafeAddDays(DateOnly date, int days)
    {
        try
        {
            return date.AddDays(days);
        }
        catch (ArgumentOutOfRangeException)
        {
            throw new ApiException(StatusCodes.Status422UnprocessableEntity, "schedule_date_out_of_range", "The schedule dates and duration exceed the supported date range.");
        }
    }

    private static bool IsWorkDay(DateOnly date, IReadOnlySet<DateOnly> holidays) =>
        date.DayOfWeek is not (DayOfWeek.Saturday or DayOfWeek.Sunday) && !holidays.Contains(date);

    private static DateOnly? Minimum(IEnumerable<DateOnly?> values) =>
        values.Where(value => value is not null).Select(value => value!.Value).Cast<DateOnly?>().Min();

    private static DateOnly? Maximum(IEnumerable<DateOnly?> values) =>
        values.Where(value => value is not null).Select(value => value!.Value).Cast<DateOnly?>().Max();
}

File: backend-node/src/schedule-calculator.ts
import { ApiError } from "./errors.js";

export type ScheduleCalculationTask = {
  id: number; parentId: number | null; sortOrder: number; planStart: string | null; planDays: number;
  startMode: string; predecessorId: number | null; lagDays: number; actualStart: string | null;
  actualFinish: string | null; forecastFinish: string | null; percentComplete: number; status: string;
};

export type ResolvedScheduleTask = {
  source: ScheduleCalculationTask; wbs: string; depth: number; planStart: string | null; planFinish: string | null;
  actualStart: string | null; actualFinish: string | null; forecastFinish: string | null; workDays: number;
  weight: number; percentComplete: number; status: string; children: ResolvedScheduleTask[];
};

function addDays(value: string, days: number): string {
  const date = new Date(`${value}T00:00:00.000Z`);
  date.setUTCDate(date.getUTCDate() + days);
  const result = date.toISOString().slice(0, 10);
  if (result < "0001-01-01" || result > "9999-12-31") throw new ApiError(422, "schedule_date_out_of_range", "The schedule dates and duration exceed the supported date range.");
  return result;
}

function isWorkDay(value: string, holidays: ReadonlySet<string>): boolean {
  const day = new Date(`${value}T00:00:00.000Z`).getUTCDay();
  return day !== 0 && day !== 6 && !holidays.has(value);
}

function nextWorkDay(value: string, holidays: ReadonlySet<string>): string {
  let candidate = value;
  while (!isWorkDay(candidate, holidays)) candidate = addDays(candidate, 1);
  return candidate;
}

export function networkDays(start: string | null, finish: string | null, holidays: ReadonlySet<string>): number {
  if (!start || !finish || finish < start) return 0;
  let count = 0;
  for (let day = start; ; day = addDays(day, 1)) {
    if (isWorkDay(day, holidays)) count += 1;
    if (day === finish) break;
  }
  return count;
}

function minimum(values: Array<string | null>): string | null {
  const available = values.filter((value): value is string => Boolean(value));
  return available.length ? available.sort()[0]! : null;
}

function maximum(values: Array<string | null>): string | null {
  const available = values.filter((value): value is string => Boolean(value));
  return available.length ? available.sort().at(-1)! : null;
}

export function resolveSchedule(tasks: ScheduleCalculationTask[], holidays: ReadonlySet<string>) {
  const bySource = new Map(tasks.map((task) => [task.id, task]));
  const children = new Map<number, ScheduleCalculationTask[]>();
  for (const task of tasks) {
    const key = task.parentId ?? 0;
    const list = children.get(key) ?? [];
    list.push(task); children.set(key, list);
  }
  for (const list of children.values()) list.sort((left, right) => left.sortOrder - right.sortOrder || left.id - right.id);
  const byId = new Map<number, ResolvedScheduleTask>();
  const visiting = new Set<number>();

  const resolveOne = (task: ScheduleCalculationTask): ResolvedScheduleTask => {
    const cached = byId.get(task.id); if (cached) return cached;
    if (visiting.has(task.id)) throw new ApiError(409, "schedule_cycle", "The schedule contains a circular parent or predecessor relationship.");
    visiting.add(task.id);
    const resolvedChildren = (children.get(task.id) ?? []).map(resolveOne);
    let result: ResolvedScheduleTask;
    if (resolvedChildren.length) {
      const planStart = minimum(resolvedChildren.map((child) => child.planStart));
      const planFinish = maximum(resolvedChildren.map((child) => child.planFinish));
      const actualStart = minimum(resolvedChildren.map((child) => child.actualStart));
      const actualFinish = resolvedChildren.every((child) => child.status === "Done") ? maximum(resolvedChildren.map((child) => child.actualFinish)) : null;
      const forecastFinish = maximum(resolvedChildren.map((child) => child.forecastFinish ?? child.planFinish));
      const weight = resolvedChildren.reduce((sum, child) => sum + child.weight, 0);
      const progress = weight ? resolvedChildren.reduce((sum, child) => sum + child.percentComplete * child.weight, 0) / weight : 0;
      const status = resolvedChildren.every((child) => child.status === "Done") ? "Done"
        : resolvedChildren.some((child) => child.status === "Blocked") ? "Blocked"
          : resolvedChildren.some((child) => child.status !== "Not Started" || child.percentComplete > 0 || child.actualStart) ? "In Progress" : "Not Started";
      result = { source: task, wbs: "", depth: 0, planStart, planFinish, actualStart, actualFinish, forecastFinish,
        workDays: networkDays(planStart, planFinish, holidays), weight: Math.max(1, weight),
        percentComplete: Math.round(progress * 100) / 100, status, children: resolvedChildren };
    } else {
      let planStart = task.planStart;
      if (task.startMode === "linked" && task.predecessorId) {
        const predecessor = bySource.get(task.predecessorId);
        if (!predecessor) throw new ApiError(409, "schedule_predecessor_missing", "A linked task refers to an inactive predecessor.");
        const predecessorFinish = resolveOne(predecessor).planFinish;
        if (predecessorFinish) planStart = nextWorkDay(addDays(predecessorFinish, 1 + task.lagDays), holidays);
      }
      const planFinish = planStart ? addDays(planStart, Math.max(1, task.planDays) - 1) : null;
      const workDays = networkDays(planStart, planFinish, holidays);
      result = { source: task, wbs: "", depth: 0, planStart, planFinish, actualStart: task.actualStart,
        actualFinish: task.actualFinish, forecastFinish: task.forecastFinish, workDays, weight: Math.max(1, workDays),
        percentComplete: task.percentComplete, status: task.status, children: [] };
    }
    visiting.delete(task.id); byId.set(task.id, result); return result;
  };

  const roots = (children.get(0) ?? []).map(resolveOne);
  tasks.forEach(resolveOne);
  const number = (siblings: ResolvedScheduleTask[], prefix: string, depth: number): void => {
    siblings.forEach((item, index) => {
      item.wbs = prefix ? `${prefix}.${index + 1}` : `${index + 1}`;
      item.depth = depth; number(item.children, item.wbs, depth + 1);
    });
  };
  number(roots, "", 0);
  return { roots, byId };
}
